File: antd-dart/example/02_data.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:antd/antd.dart';

/// Demonstrates storing and retrieving public immutable data.
void main() async {
  final client = AntdClient();

  try {
    // Store data
    final data = Uint8List.fromList(utf8.encode('Hello, Autonomi!'));
    final result = await client.dataPutPublic(data);
    print('Stored at ${result.address} (cost: ${result.cost} atto)');

    // Retrieve data
    final retrieved = await client.dataGetPublic(result.address);
    print('Retrieved: ${utf8.decode(retrieved)}');

    // Estimate cost
    final cost = await client.dataCost(data);
    print('Estimated cost: $cost atto');
  } on AntdError catch (e) {
    print('Error: $e');
  } finally {
    client.close();
  }
}

File: antd-cpp/examples/02-data.cpp
#include <iostream>
#include <string>
#include <vector>
#include "antd/antd.hpp"

int main() {
    try {
        antd::Client client;

        // Store public data
        std::string msg = "Hello, Autonomi!";
        std::vector<uint8_t> data(msg.begin(), msg.end());

        auto result = client.data_put_public(data);
        std::cout << "Stored at: " << result.address << "\n";
        std::cout << "Cost: " << result.cost << " atto\n";

        // Retrieve it
        auto retrieved = client.data_get_public(result.address);
        std::string text(retrieved.begin(), retrieved.end());
        std::cout << "Retrieved: " << text << "\n";

        // Estimate cost
        auto cost = client.data_cost(data);
        std::cout << "Estimated cost: " << cost << " atto\n";

    } catch (const antd::AntdError& e) {
        std::cerr << "Error: " << e.what() << "\n";
        return 1;
    }
    return 0;
}
